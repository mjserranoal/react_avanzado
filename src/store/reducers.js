import {
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT_SUCCESS,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_CREATED_SUCCESS,
  ADVERT_LOADED_SUCCESS,
  UI_RESET_ERROR,
  ADVERT_DELETED_SUCCESS,
} from './types';

export const defaultState = {
  auth: false,
  adverts: {
    areLoaded: false,
    data: [],
  },
  ui: {
    isLoading: false,
    error: null,
  },
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCESS:
      return true;
    case AUTH_LOGOUT_SUCCESS:
      return false;
    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  if (action.type === ADVERTS_LOADED_SUCCESS) {
    return { areLoaded: true, data: action.payload };
  }
  if (action.type === ADVERT_LOADED_SUCCESS) {
    return { ...state, data: [action.payload] };
  }
  if (action.type === ADVERT_CREATED_SUCCESS) {
    return { ...state, data: [action.payload].concat(state.data) };
  }
  if (action.type === ADVERT_DELETED_SUCCESS) {
    return {
      ...state,
      data: state.data.filter(advert => advert.id !== action.payload),
    };
  }

  return state;
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return { isLoading: false, error: action.payload };
  }

  if (/_REQUEST$/.test(action.type)) {
    return { isLoading: true, error: null };
  }

  if (/_SUCCESS$/.test(action.type)) {
    return { isLoading: false, error: null };
  }

  if (action.type === UI_RESET_ERROR) {
    return { ...state, error: null };
  }

  return state;
}
