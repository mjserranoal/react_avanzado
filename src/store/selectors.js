export const getIsLogged = state => state.auth;

export const getAdverts = state =>
  state.adverts.areLoaded ? state.adverts.data : [];

export const getAdvert = advertId => state => {
  const advert = state.adverts.data.find(advert => advert.id === advertId);
  return advert;
};

export const getUi = state => state.ui;

export const areAdvertsLoaded = state => state.adverts.areLoaded;
