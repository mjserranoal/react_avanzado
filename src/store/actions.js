import { areAdvertsLoaded, getAdvert } from './selectors';
import {
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_LOADED_FAILURE,
  ADVERT_LOADED_SUCCESS,
  ADVERT_LOADED_REQUEST,
  UI_RESET_ERROR,
  ADVERT_CREATED_SUCCESS,
  ADVERT_CREATED_REQUEST,
  ADVERT_CREATED_FAILURE,
  AUTH_LOGOUT_FAILURE,
  AUTH_LOGOUT_SUCCESS,
  AUTH_LOGOUT_REQUEST,
  ADVERT_DELETED_SUCCESS,
  ADVERT_DELETED_FAILURE,
} from './types';

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});

export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  error: true,
  payload: error,
});

export const authLogin =
  credentials =>
  async (dispatch, _getState, { service, router }) => {
    dispatch(authLoginRequest());
    try {
      await service.auth.login(credentials);
      // Logged in
      dispatch(authLoginSuccess());
      // Redirect to pathname
      const to = router.state.location.state?.from?.pathname || '/';
      router.navigate(to);
    } catch (error) {
      dispatch(authLoginFailure(error));
    }
  };

export const authLogoutRequest = () => ({
  type: AUTH_LOGOUT_REQUEST,
});

export const authLogoutSuccess = () => ({
  type: AUTH_LOGOUT_SUCCESS,
});

export const authLogoutFailure = error => ({
  type: AUTH_LOGOUT_FAILURE,
  error: true,
  payload: error,
});

export const authLogout =
  () =>
  async (dispatch, _getState, { service }) => {
    dispatch(authLogoutRequest());
    try {
      await service.auth.logout();
      dispatch(authLogoutSuccess());
    } catch (error) {
      dispatch(authLogoutFailure(error));
    }
  };

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = adverts => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});

export const advertsLoadedFailure = error => ({
  type: ADVERTS_LOADED_FAILURE,
  error: true,
  payload: error,
});

export const advertsLoaded =
  () =>
  async (dispatch, getState, { service }) => {
    if (areAdvertsLoaded(getState())) {
      return;
    }
    dispatch(advertsLoadedRequest());
    try {
      const adverts = await service.adverts.getLatestAdverts();
      adverts.forEach(advert => {
        if (!('photo' in advert)) {
          advert.photo = false;
        }
      });
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
    }
  };

export const advertLoadedRequest = () => ({
  type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedSuccess = advert => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: advert,
});

export const advertLoadedFailure = error => ({
  type: ADVERT_LOADED_FAILURE,
  error: true,
  payload: error,
});

export const advertLoad =
  advertId =>
  async (dispatch, getState, { service }) => {
    const isLoaded = getAdvert(advertId)(getState());
    if (isLoaded) {
      return;
    }
    dispatch(advertLoadedRequest());
    try {
      const advert = await service.adverts.getAdvert(advertId);
      dispatch(advertLoadedSuccess(advert));
    } catch (error) {
      dispatch(advertLoadedFailure(error));
    }
  };

export const advertCreateRequest = () => ({
  type: ADVERT_CREATED_REQUEST,
});

export const advertCreateSuccess = advert => ({
  type: ADVERT_CREATED_SUCCESS,
  payload: advert,
});

export const advertCreateFailure = error => ({
  type: ADVERT_CREATED_FAILURE,
  error: true,
  payload: error,
});

export const advertDeleteRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertDeleteSuccess = advertId => ({
  type: ADVERT_DELETED_SUCCESS,
  payload: advertId,
});

export const advertDeleteFailure = error => ({
  type: ADVERT_DELETED_FAILURE,
  error: true,
  payload: error,
});

export const advertDelete =
  advertId =>
  async (dispatch, _getState, { service, router }) => {
    dispatch(advertDeleteRequest());
    try {
      await service.adverts.deleteAdvert(advertId);
      dispatch(advertDeleteSuccess(advertId));
      router.navigate(`/adverts`);
    } catch (error) {
      dispatch(advertCreateFailure(error));
    }
  };

export const advertCreate =
  advert =>
  async (dispatch, _getState, { service, router }) => {
    dispatch(advertCreateRequest());
    try {
      const { id } = await service.adverts.createAdvert(advert);
      const createdAdvert = await service.adverts.getAdvert(id);
      dispatch(advertCreateSuccess(createdAdvert));
      router.navigate(`/adverts/${id}`);
      return createdAdvert;
    } catch (error) {
      dispatch(advertCreateFailure(error));
    }
  };

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
