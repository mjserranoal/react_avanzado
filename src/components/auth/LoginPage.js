import { useState } from 'react';

import Button from '../shared/Button';
import FormField from '../shared/FormField';

import './LoginPage.css';
import { useDispatch, useSelector } from 'react-redux';
import { authLogin, uiResetError } from '../../store/actions';
import { getUi } from '../../store/selectors';

function LoginPage() {
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);
  const [credentials, setCredentials] = useState({
    email: '',
    password: '',
    remember: false,
  });

  const resetError = () => {
    dispatch(uiResetError());
  };

  const handleSubmit = async event => {
    event.preventDefault();
    dispatch(authLogin(credentials));
  };

  const handleChange = event => {
    let valor = event.target.value;

    if (event.target.name === 'remember') {
      valor = event.target.checked;
    }

    setCredentials({
      ...credentials,
      [event.target.name]: valor,
    });
  };

  const buttonDisabled =
    isLoading || !credentials.email || !credentials.password;

  return (
    <div className="loginPage">
      <h1 className="loginPage-title">Log in to Nodepop</h1>
      <form onSubmit={handleSubmit}>
        <FormField
          type="text"
          name="email"
          label="email"
          className="loginForm-field"
          onChange={handleChange}
          value={credentials.email}
          autofocus
        />
        <FormField
          type="password"
          name="password"
          label="password"
          className="loginForm-field"
          onChange={handleChange}
          value={credentials.password}
        />
        <FormField
          type="checkbox"
          name="remember"
          label="remember me!"
          className="loginForm-field"
          onChange={handleChange}
          checked={credentials.remember}
        />
        <Button
          type="submit"
          variant="primary"
          className="loginForm-submit"
          disabled={buttonDisabled}
        >
          Log in
        </Button>
      </form>
      {error && (
        <div onClick={resetError} className="loginPage-error">
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
