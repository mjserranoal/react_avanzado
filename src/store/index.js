import { createStore, combineReducers, applyMiddleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';

import * as service from '../service';

// import { auth, tweets } from './reducers';
import * as reducers from './reducers';
import * as actionCreators from './actions';

// const reducer = combineReducers({
//   auth,
//   tweets,
// });

const reducer = combineReducers(reducers);

const composeEnhancers = composeWithDevTools({
  actionCreators,
});

// const thunk = store => next => action => {}

// const thunk = function (store) {
//   return function (next) {
//     return function (action) {
//       if (typeof action === 'function') {
//         return action(store.dispatch, store.getState, extraArgument);
//       }
//       return next(action);
//     };
//   };
// };

const logger = store => next => action => {
  if (action.type) {
    console.group(action.type);
    console.info('dispatching', action, store.getState());
    const result = next(action);
    console.log('next state', store.getState());
    console.groupEnd();
    return result;
  }
  return next(action);
};

const timestamp = () => next => action => {
  return next({
    ...action,
    meta: {
      ...action.meta,
      timestamp: Date.now(),
    },
  });
};

const failureRedirects = (router, redirectsMap) => () => next => action => {
  const result = next(action);

  if (action.error) {
    const redirect = redirectsMap[action.payload.status];
    if (redirect) {
      router.navigate(redirect);
    }
  }
  return result;
};

export default function configureStore(preloadedState, { router }) {
  const middleware = [
    thunk.withExtraArgument({ service, router }),
    timestamp,
    failureRedirects(router, { 401: '/login', 404: '/404' }),
    logger,
  ];

  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(applyMiddleware(...middleware)),
  );
  return store;
}
