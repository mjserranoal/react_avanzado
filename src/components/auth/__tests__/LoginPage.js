import { render, screen } from '@testing-library/react';
import LoginPage from '../LoginPage';
import { Provider } from 'react-redux';
import { defaultState } from '../../../store/reducers';
import { authLogin, uiResetError } from '../../../store/actions';
import userEvent from '@testing-library/user-event';

jest.mock('../../../store/actions');

describe('LoginPage', () => {
  const renderComponent = (error = null) => {
    const store = {
      getState: () => {
        const state = defaultState;
        state.ui.error = error;

        return state;
      },
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );
  };

  test('snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('should dispatch authLogin action', () => {
    const email = 'mj@mj.com';
    const password = '1234';
    const remember = false;

    renderComponent();

    const emailInput = screen.getByLabelText(/email/);
    const passwordInput = screen.getByLabelText(/password/);
    const submitButton = screen.getByRole('button');

    expect(submitButton).toBeDisabled();

    userEvent.type(emailInput, email);
    userEvent.type(passwordInput, password);

    expect(submitButton).toBeEnabled();

    userEvent.click(submitButton);

    expect(authLogin).toHaveBeenCalledWith({ email, password, remember });
  });

  test('should display an error', () => {
    const error = { message: 'Unauthorized' };
    renderComponent(error);

    const errorElement = screen.getByText(error.message);
    expect(errorElement).toBeInTheDocument();

    userEvent.click(errorElement);

    expect(uiResetError).toHaveBeenCalled();
  });
});
