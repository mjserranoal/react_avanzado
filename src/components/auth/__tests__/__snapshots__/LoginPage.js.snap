// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in to Nodepop
    </h1>
    <form>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            email
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            password
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            remember me!
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="remember"
            type="checkbox"
          />
        </label>
      </div>
      <button
        class="sc-beqWaB dChUv loginForm-submit"
        disabled=""
        type="submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;
